### ISSUE TITLE
libevent-2.0.22-stable make check bufferevent_timeout FAILED

### ISSUE THREAD

**Issue by dregerATiTD** [libevent/libevent#262]
I get 1 failure in "make check", follows is the output of make check and ./test/regress.

Running tests:
EVPORT
Skipping test
KQUEUE
Skipping test
EPOLL
 test-eof: OKAY
 test-weof: OKAY
 test-time: OKAY
 test-changelist: OKAY
 regress: bufferevent/bufferevent_timeout:
  FAIL regress_bufferevent.c:768: want(res1.n_write_timeouts)
  FAIL regress_bufferevent.c:770: want(res1.n_write_timeouts == 1)
  FAIL regress_bufferevent.c:773: assert(abs(timeval_msec_diff(((&started_at)), ((&res1.write_timeout_at))) - (100)) <= 50): 2009747997 vs 50
  [bufferevent_timeout FAILED]
1/172 TESTS FAILED. (0 skipped)
FAILED
EPOLL (changelist)
 test-eof: OKAY
 test-weof: OKAY
 test-time: OKAY
 test-changelist: OKAY
 regress: bufferevent/bufferevent_timeout:
  FAIL regress_bufferevent.c:768: want(res1.n_write_timeouts)
  FAIL regress_bufferevent.c:770: want(res1.n_write_timeouts == 1)
  FAIL regress_bufferevent.c:773: assert(abs(timeval_msec_diff(((&started_at)), ((&res1.write_timeout_at))) - (100)) <= 50): 2009691103 vs 50
  [bufferevent_timeout FAILED]
1/172 TESTS FAILED. (0 skipped)
FAILED
DEVPOLL
Skipping test
POLL
 test-eof: OKAY
 test-weof: OKAY
 test-time: OKAY
 test-changelist: OKAY
 regress: bufferevent/bufferevent_timeout:
  FAIL regress_bufferevent.c:768: want(res1.n_write_timeouts)
  FAIL regress_bufferevent.c:770: want(res1.n_write_timeouts == 1)
  FAIL regress_bufferevent.c:773: assert(abs(timeval_msec_diff(((&started_at)), ((&res1.write_timeout_at))) - (100)) <= 50): 2009634194 vs 50
  [bufferevent_timeout FAILED]
1/172 TESTS FAILED. (0 skipped)
FAILED
SELECT
 test-eof: OKAY
 test-weof: OKAY
 test-time: OKAY
 test-changelist: OKAY
 regress: bufferevent/bufferevent_timeout:
  FAIL regress_bufferevent.c:768: want(res1.n_write_timeouts)
  FAIL regress_bufferevent.c:770: want(res1.n_write_timeouts == 1)
  FAIL regress_bufferevent.c:773: assert(abs(timeval_msec_diff(((&started_at)), ((&res1.write_timeout_at))) - (100)) <= 50): 2009577345 vs 50
  [bufferevent_timeout FAILED]
1/172 TESTS FAILED. (0 skipped)
FAILED
WIN32
Skipping test
# FAIL: test-script.sh
#1 of 1 test failed

[root@xxxxxxxx libevent-2.0.22-stable]# ./test/regress
main/methods: [forking] OK
main/version: OK
main/base_features: [forking] OK
main/base_environ: [forking] OK
main/event_base_new: [forking] OK
main/free_active_base: [forking] OK
main/manipulate_active_events: [forking] OK
main/bad_assign: [forking] OK
main/bad_reentrant: [forking] OK
main/persistent_timeout: [forking] OK
main/persistent_timeout_jump: [forking] OK
main/persistent_active_timeout: [forking] OK
main/priorities: [forking] OK
main/priority_active_inversion: [forking] OK
main/common_timeout: [forking] OK
main/simpleread: [forking] OK
main/simpleread_multiple: [forking] OK
main/simplewrite: [forking] OK
main/simpleclose: [forking] OK
main/multiple: [forking] OK
main/persistent: [forking] OK
main/combined: [forking] OK
main/simpletimeout: [forking] OK
main/loopbreak: [forking] OK
main/loopexit: [forking] OK
main/loopexit_multiple: [forking] OK
main/nonpersist_readd: [forking] OK
main/multiple_events_for_same_fd: [forking] OK
main/want_only_once: [forking] OK
main/event_once: [forking] OK
main/event_pending: [forking] OK
main/dup_fd: [forking] OK
main/mm_functions: [forking] OK
main/many_events: [forking] OK
main/many_events_slow_add: [forking] OK
main/struct_event_size: OK
main/fork: [forking] OK
heap/randomized: OK
et/et: [forking] OK
et/et_mix_error: [forking] OK
evbuffer/evbuffer: OK
evbuffer/remove_buffer_with_empty: OK
evbuffer/reserve2: OK
evbuffer/reserve_many: OK
evbuffer/reserve_many2: OK
evbuffer/reserve_many3: OK
evbuffer/expand: OK
evbuffer/reference: OK
evbuffer/iterative: OK
evbuffer/readln: OK
evbuffer/search_eol: OK
evbuffer/find: OK
evbuffer/ptr_set: OK
evbuffer/search: OK
evbuffer/callbacks: OK
evbuffer/add_reference: OK
evbuffer/prepend: [forking] OK
evbuffer/peek: OK
evbuffer/freeze_start: OK
evbuffer/freeze_end: OK
evbuffer/add_file_sendfile: [forking] OK
evbuffer/add_file_mmap: [forking] OK
evbuffer/add_file_linear: [forking] OK
signal/simplesignal: [forking] OK
signal/multiplesignal: [forking] OK
signal/immediatesignal: [forking] OK
signal/signal_dealloc: [forking] OK
signal/signal_pipeloss: [forking] OK
signal/signal_switchbase: [forking] OK
signal/signal_restore: [forking] OK
signal/signal_assert: [forking] OK
signal/signal_while_processing: [forking] OK
util/ipv4_parse: OK
util/ipv6_parse: OK
util/sockaddr_port_parse: OK
util/sockaddr_port_format: OK
util/sockaddr_predicates: OK
util/evutil_snprintf: OK
util/evutil_strtoll: OK
util/evutil_casecmp: OK
util/strlcpy: OK
util/log: [forking] OK
util/upcast: OK
util/integers: OK
util/rand: [forking] OK
util/getaddrinfo: [forking] OK
bufferevent/bufferevent: [forking] OK
bufferevent/bufferevent_pair: [forking] OK
bufferevent/bufferevent_watermarks: [forking] OK
bufferevent/bufferevent_pair_watermarks: [forking] OK
bufferevent/bufferevent_filters: [forking] OK
bufferevent/bufferevent_pair_filters: [forking] OK
bufferevent/bufferevent_connect: [forking] OK
bufferevent/bufferevent_connect_defer: [forking] OK
bufferevent/bufferevent_connect_lock: [forking] OK
bufferevent/bufferevent_connect_lock_defer: [forking] OK
bufferevent/bufferevent_connect_unlocked_cbs: [forking] OK
bufferevent/bufferevent_connect_fail: [forking] OK
bufferevent/bufferevent_timeout: [forking]
  FAIL regress_bufferevent.c:768: want(res1.n_write_timeouts)
  FAIL regress_bufferevent.c:770: want(res1.n_write_timeouts == 1)
  FAIL regress_bufferevent.c:773: assert(abs(timeval_msec_diff(((&started_at)), ((&res1.write_timeout_at))) - (100)) <= 50): 2009181826 vs 50
  [bufferevent_timeout FAILED]
bufferevent/bufferevent_timeout_pair: [forking] OK
bufferevent/bufferevent_timeout_filter: [forking] OK
bufferevent/bufferevent_timeout_filter_pair: [forking] OK
bufferevent/bufferevent_zlib: [forking] OK
http/primitives: OK
http/base: [forking] OK
http/bad_headers: OK
http/parse_query: OK
http/parse_uri: OK
http/parse_uri_nc: OK
http/uriencode: OK
http/basic: [forking] OK
http/cancel: [forking] OK
http/virtual_host: [forking] OK
http/post: [forking] OK
http/put: [forking] OK
http/delete: [forking] OK
http/allowed_methods: [forking] OK
http/failure: [forking] OK
http/connection: [forking] OK
http/persist_connection: [forking] OK
http/connection_async: [forking] OK
http/close_detection: [forking] OK
http/close_detection_delay: [forking] OK
http/bad_request: [forking] OK
http/incomplete: [forking] OK
http/incomplete_timeout: [forking] OK
http/terminate_chunked: [forking] OK
http/highport: [forking] OK
http/dispatcher: [forking] OK
http/multi_line_header: [forking] OK
http/negative_content_length: [forking] OK
http/chunk_out: [forking] OK
http/stream_out: [forking] OK
http/stream_in: [forking] OK
http/stream_in_cancel: [forking] OK
http/connection_fail: [forking] OK
http/connection_retry: [forking] OK
http/data_length_constraints: [forking] OK
dns/server: [forking] OK
dns/gethostbyname: [forking] OK
dns/gethostbyname6: [forking] OK
dns/gethostbyaddr: [forking] OK
dns/resolve_reverse: [forking] OK
dns/search: [forking] OK
dns/search_cancel: [forking] OK
dns/retry: [forking] OK
dns/reissue: [forking] OK
dns/inflight: [forking] OK
dns/bufferevent_connect_hostname: [forking] OK
dns/getaddrinfo_async: [forking] OK
dns/getaddrinfo_cancel_stress: [forking] OK
dns/leak_shutdown: [forking] OK
dns/leak_cancel: [forking] OK
evtag/int: [forking] OK
evtag/fuzz: [forking] OK
evtag/encoding: [forking] OK
evtag/peek: OK
rpc/basic_test: [forking] OK
rpc/basic_message: [forking] OK
rpc/basic_client: [forking] OK
rpc/basic_queued_client: [forking] OK
rpc/basic_client_with_pause: [forking] OK
rpc/client_timeout: [forking] OK
rpc/test: [forking] OK
thread/basic: [forking] OK
thread/forking: [forking] OK
thread/conditions_simple: [forking] OK
thread/deferred_cb_skew: [forking] OK
listener/randport: [forking] OK
listener/randport_ts: [forking] OK
listener/error: [forking] OK
listener/error_ts: [forking] OK
1/172 TESTS FAILED. (0 skipped)

**Comment by ellzey** [comment]
Yes, I have noted this test. Stupid question: is this on a VM?

**Comment by dregerATiTD** [comment]
No, I wasn't using a VM.  Sorry its been such a long time in responding.

**Comment by code-share1** [comment]
When run regress test, I encounter below error:
bufferevent/bufferevent_timeout_filter: [forking] 
  FAIL ../../libevent-2.0.22-stable/test/regress_bufferevent.c:772: assert(abs(timeval_msec_diff(((&started_at)), ((&res1.read_timeout_at))) - (150)) <= 50): 132 vs 50
  [bufferevent_timeout_filter FAILED]
bufferevent/bufferevent_timeout_filter_pair: [forking] OK

The full regress test result as:
main/methods: [forking] OK
main/version: OK
main/base_features: [forking] OK
main/base_environ: [forking] OK
main/event_base_new: [forking] OK
main/free_active_base: [forking] OK
main/manipulate_active_events: [forking] OK
main/bad_assign: [forking] OK
main/bad_reentrant: [forking] OK
main/persistent_timeout: [forking] OK
main/persistent_timeout_jump: [forking] OK
main/persistent_active_timeout: [forking] OK
main/priorities: [forking] OK
main/priority_active_inversion: [forking] OK
main/common_timeout: [forking] OK
main/simpleread: [forking] OK
main/simpleread_multiple: [forking] OK
main/simplewrite: [forking] OK
main/simpleclose: [forking] OK
main/multiple: [forking] OK
main/persistent: [forking] OK
main/combined: [forking] OK
main/simpletimeout: [forking] OK
main/loopbreak: [forking] OK
main/loopexit: [forking] OK
main/loopexit_multiple: [forking] OK
main/nonpersist_readd: [forking] OK
main/multiple_events_for_same_fd: [forking] OK
main/want_only_once: [forking] OK
main/event_once: [forking] OK
main/event_pending: [forking] OK
main/dup_fd: [forking] OK
main/mm_functions: [forking] OK
main/many_events: [forking] OK
main/many_events_slow_add: [forking] OK
main/struct_event_size: OK
main/fork: [forking] OK
heap/randomized: OK
et/et: [forking] OK
et/et_mix_error: [forking] OK
evbuffer/evbuffer: OK
evbuffer/remove_buffer_with_empty: OK
evbuffer/reserve2: OK
evbuffer/reserve_many: OK
evbuffer/reserve_many2: OK
evbuffer/reserve_many3: OK
evbuffer/expand: OK
evbuffer/reference: OK
evbuffer/iterative: OK
evbuffer/readln: OK
evbuffer/search_eol: OK
evbuffer/find: OK
evbuffer/ptr_set: OK
evbuffer/search: OK
evbuffer/callbacks: OK
evbuffer/add_reference: OK
evbuffer/prepend: [forking] OK
evbuffer/peek: OK
evbuffer/freeze_start: OK
evbuffer/freeze_end: OK
evbuffer/add_file_sendfile: [forking] OK
evbuffer/add_file_mmap: [forking] OK
evbuffer/add_file_linear: [forking] OK
signal/simplesignal: [forking] OK
signal/multiplesignal: [forking] OK
signal/immediatesignal: [forking] OK
signal/signal_dealloc: [forking] OK
signal/signal_pipeloss: [forking] OK
signal/signal_switchbase: [forking] OK
signal/signal_restore: [forking] OK
signal/signal_assert: [forking] OK
signal/signal_while_processing: [forking] OK
util/ipv4_parse: OK
util/ipv6_parse: OK
util/sockaddr_port_parse: OK
util/sockaddr_port_format: OK
util/sockaddr_predicates: OK
util/evutil_snprintf: OK
util/evutil_strtoll: OK
util/evutil_casecmp: OK
util/strlcpy: OK
util/log: [forking] OK
util/upcast: OK
util/integers: OK
util/rand: [forking] OK
util/getaddrinfo: [forking] OK
bufferevent/bufferevent: [forking] OK
bufferevent/bufferevent_pair: [forking] OK
bufferevent/bufferevent_watermarks: [forking] OK
bufferevent/bufferevent_pair_watermarks: [forking] OK
bufferevent/bufferevent_filters: [forking] OK
bufferevent/bufferevent_pair_filters: [forking] OK
bufferevent/bufferevent_connect: [forking] OK
bufferevent/bufferevent_connect_defer: [forking] OK
bufferevent/bufferevent_connect_lock: [forking] OK
bufferevent/bufferevent_connect_lock_defer: [forking] OK
bufferevent/bufferevent_connect_unlocked_cbs: [forking] OK
bufferevent/bufferevent_connect_fail: [forking] OK
bufferevent/bufferevent_timeout: [forking] OK
bufferevent/bufferevent_timeout_pair: [forking] OK
bufferevent/bufferevent_timeout_filter: [forking] 
  FAIL ../../libevent-2.0.22-stable/test/regress_bufferevent.c:772: assert(abs(timeval_msec_diff(((&started_at)), ((&res1.read_timeout_at))) - (150)) <= 50): 132 vs 50
  [bufferevent_timeout_filter FAILED]
bufferevent/bufferevent_timeout_filter_pair: [forking] OK
bufferevent/bufferevent_zlib: [forking] OK
http/primitives: OK
http/base: [forking] OK
http/bad_headers: OK
http/parse_query: OK
http/parse_uri: OK
http/parse_uri_nc: OK
http/uriencode: OK
http/basic: [forking] OK
http/cancel: [forking] OK
http/virtual_host: [forking] OK
http/post: [forking] OK
http/put: [forking] OK
http/delete: [forking] OK
http/allowed_methods: [forking] OK
http/failure: [forking] OK
http/connection: [forking] OK
http/persist_connection: [forking] OK
http/connection_async: [forking] OK
http/close_detection: [forking] OK
http/close_detection_delay: [forking] OK
http/bad_request: [forking] OK
http/incomplete: [forking] OK
http/incomplete_timeout: [forking] OK
http/terminate_chunked: [forking] OK
http/highport: [forking] OK
http/dispatcher: [forking] OK
http/multi_line_header: [forking] OK
http/negative_content_length: [forking] OK
http/chunk_out: [forking] OK
http/stream_out: [forking] OK
http/stream_in: [forking] OK
http/stream_in_cancel: [forking] OK
http/connection_fail: [forking] OK
http/connection_retry: [forking] OK
http/data_length_constraints: [forking] OK
dns/server: [forking] OK
dns/gethostbyname: [forking] OK
dns/gethostbyname6: [forking] OK
dns/gethostbyaddr: [forking] OK
dns/resolve_reverse: [forking] OK
dns/search: [forking] OK
dns/search_cancel: [forking] OK
dns/retry: [forking] OK
dns/reissue: [forking] OK
dns/inflight: [forking] OK
dns/bufferevent_connect_hostname: [forking] OK
dns/getaddrinfo_async: [forking] OK
dns/getaddrinfo_cancel_stress: [forking] OK
dns/leak_shutdown: [forking] OK
dns/leak_cancel: [forking] OK
evtag/int: [forking] OK
evtag/fuzz: [forking] OK
evtag/encoding: [forking] OK
evtag/peek: OK
rpc/basic_test: [forking] OK
rpc/basic_message: [forking] OK
rpc/basic_client: [forking] OK
rpc/basic_queued_client: [forking] OK
rpc/basic_client_with_pause: [forking] OK
rpc/client_timeout: [forking] OK
rpc/test: [forking] OK
thread/basic: [forking] OK
thread/forking: [forking] OK
thread/conditions_simple: [forking] OK
thread/deferred_cb_skew: [forking] OK
listener/randport: [forking] OK
listener/randport_ts: [forking] OK
listener/error: [forking] OK
listener/error_ts: [forking] OK
1/172 TESTS FAILED. (0 skipped)

**Comment by azat** [comment]
This is because of timers issues, i.e. when computer is slow then usual you can get this errors.

**Comment by code-share1** [comment]
Thanks azat's response! Then how can I remove this error?

**Comment by azat** [comment]
You can increase diff there, i.e. `50` to `200`

P.S. of course this should be fixed in the long-term

**Comment by code-share1** [comment]
Thanks azat!
I ever considered to increase the tolerance too. But I'm just confusing what's the reasonable tolerance value.

test_timeval_diff_eq(&started_at, &res1.read_timeout_at, 150);
->
#define test_timeval_diff_eq(tv1, tv2, diff)                            \                                                                     
        test_timeval_diff_leq((tv1), (tv2), (diff), 50)
->
#define test_timeval_diff_leq(tv1, tv2, diff, tolerance)                \
        tt_int_op(abs(timeval_msec_diff((tv1), (tv2)) - diff), <=, tolerance)

**Comment by awilfox** [comment]
I'm seeing this exact failure mode with libevent 2.1.8, on a 96-core physical ARM 64-bit server.  There is no VM or hypervisor.  The failure is very fast (I would say within 5 seconds of the test starting).

```
bufferevent/bufferevent_timeout: 
  FAIL test/regress_bufferevent.c:1035: want(res1.n_write_timeouts)
  FAIL test/regress_bufferevent.c:1037: want(res1.n_write_timeouts == 1)
  FAIL test/regress_bufferevent.c:1040: assert(labs(timeval_msec_diff(((&started_at)), ((&res1.write_timeout_at))) - (100)) <= 50): 1527483315113 vs 50
  [bufferevent_timeout FAILED]
bufferevent/bufferevent_timeout: 
  FAIL test/regress_bufferevent.c:1035: want(res1.n_write_timeouts)
  FAIL test/regress_bufferevent.c:1037: want(res1.n_write_timeouts == 1)
  FAIL test/regress_bufferevent.c:1040: assert(labs(timeval_msec_diff(((&started_at)), ((&res1.write_timeout_at))) - (100)) <= 50): 1527483315169 vs 50
  [bufferevent_timeout FAILED]
bufferevent/bufferevent_timeout: 
  FAIL test/regress_bufferevent.c:1035: want(res1.n_write_timeouts)
  FAIL test/regress_bufferevent.c:1037: want(res1.n_write_timeouts == 1)
  FAIL test/regress_bufferevent.c:1040: assert(labs(timeval_msec_diff(((&started_at)), ((&res1.write_timeout_at))) - (100)) <= 50): 1527483315172 vs 50
  [bufferevent_timeout FAILED]
bufferevent/bufferevent_timeout: 
  FAIL test/regress_bufferevent.c:1035: want(res1.n_write_timeouts)
  FAIL test/regress_bufferevent.c:1037: want(res1.n_write_timeouts == 1)
  FAIL test/regress_bufferevent.c:1040: assert(labs(timeval_msec_diff(((&started_at)), ((&res1.write_timeout_at))) - (100)) <= 50): 1527483315198 vs 50
  [bufferevent_timeout FAILED]
make[3]: *** [Makefile:2715: test_runner_timerfd] Error 1
1/312 TESTS FAILED. (17 skipped)
FAILED
make[3]: *** [Makefile:2703: test_runner_select] Error 1
1/312 TESTS FAILED. (17 skipped)
FAILED
make[3]: *** [Makefile:2717: test_runner_changelist] Error 1
1/312 TESTS FAILED. (17 skipped)
FAILED
make[3]: *** [Makefile:2719: test_runner_timerfd_changelist] Error 1
1/312 TESTS FAILED. (17 skipped)
FAILED
```

**Comment by azat** [comment]
On Thu, May 31, 2018 at 05:43:54PM -0700, A. Wilcox wrote:
> I'm seeing this exact failure mode with libevent 2.1.8, on a 96-core physical ARM 64-bit server.  There is no VM or hypervisor.  The failure is very fast (I would say within 5 seconds of the test starting).

Thanks for reporting this!
Hm, does your server under high load?

Anyway I have some ideas how to rewrite it, to reduce race window (I
started doing this in 65eca09fe4422fbfb4ce3b254711597d9e3b3794 and
602455bce023d07b11554699e318724564b00151 for thread/no_events)

If you are curious here is the reproducer (you should be in built
directory for this of course):
  $ docker run --cpus=0.01 -e LD_LIBRARY_PATH=$PWD/lib -e PATH=/usr/bin:/bin:$PWD/bin -v $PWD:$PWD --name le --rm -it debian:testing regress --no-fork --verbose bufferevent/bufferevent_timeout
(or you can do the same with `systemd-run -p CPUQuota=1%`)

**Comment by awilfox** [comment]
Only one package builds at a time on these servers, as they are dedicated package building machines.  The only load would be the other tests running.

The system in question does not run systemd and does not have Docker available.

**Comment by azat** [comment]
> Only one package builds at a time on these servers, as they are dedicated package building machines.  The only load would be the other tests running.

libevent tests?
Also can you verify with EVENT_PRECISE_TIMER=1 set (just env variable) ?

**Comment by azat** [comment]
@awilfox forget what I asked. I finally fixed this test (well not *1000%* but 99% of the time it should pass). It has been fixed in the master, but will be backported to the 2.0 branch too.

**Comment by azat** [comment]
The backport landed as 9e6f650
